File: testbed_php/test/associations/ManyToMStarTest.php
<?php

class ManyToMStarTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentR(99);
    $this->assertEqual(0,$s->numberOfMentors());
  }
  
  public function test_ConstructorWithTooFew()
  {
  	$this->expectException(new Exception("Unable to create MentorR, must have at least 3 students"));
    $s = new StudentR(99);
    new MentorR("blah",array($s));
  }


  public function test_ConstructorCannotHaveTooMany()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    $s5 = new StudentR(95);
    $s6 = new StudentR(94);
    $s7 = new StudentR(93);
    $s8 = new StudentR(92);
    new MentorR("blah",array($s,$s2,$s3,$s4,$s5,$s6,$s7,$s8));
  }
  
  

  public function test_constructorJustBigEnough()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);

    $m = new MentorR("blah",array($s,$s2,$s3));

    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
  }
  

  public function test_constructorMoreThanBigEnough()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);

    $m = new MentorR("blah",array($s,$s2,$s3,$s4));

    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));
  }  


  public function test_constructorWatchOutForDuplicateEntries()
  {
  	$this->expectException(new Exception("Unable to create MentorR, must have at least 3 students"));
    $s = new StudentR(99);
    new MentorR("blah",array($s,$s));
  }
  

  public function test_constructorExistingMentorsOkay()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    
    $m = new MentorR("blah",array($s,$s2,$s3));

    $this->assertEqual(3, $m->numberOfStudents());
    $this->assertEqual($m, $s->getMentor(0));
    $this->assertEqual($m, $s2->getMentor(0));
    $this->assertEqual($m, $s3->getMentor(0));
    $this->assertEqual(0, $s4->numberOfMentors());    

    $m2 = new MentorR("blah",array($s,$s2,$s4));
    
    $this->assertEqual(3, $m->numberOfStudents());
    $this->assertEqual($m, $s->getMentor(0));
    $this->assertEqual($m, $s2->getMentor(0));
    $this->assertEqual($m, $s3->getMentor(0));
    
    $this->assertEqual(3, $m2->numberOfStudents());
    $this->assertEqual($m2, $s->getMentor(1));
    $this->assertEqual($m2, $s2->getMentor(1));
    $this->assertEqual($m2, $s4->getMentor(0));

  }   
  

  public function test_setStudents_tooFew()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    $s5 = new StudentR(95);

    $m = new MentorR("blah",array($s,$s2,$s3));
    
    $this->assertEqual(false, $m->setStudents(array($s5,$s4)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
    
    
  }


  public function test_setStudents_doNotAllowDuplicates()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    
    $m = new MentorR("blah",array($s,$s2,$s3));
    $this->assertEqual(false, $m->setStudents($s2,$s2,$s3));
  }


  public function test_setStudents_aboveMinimum()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);

    $m = new MentorR("blah",array($s,$s2,$s3));
    
    $this->assertEqual(true, $m->setStudents(array($s2,$s3,$s4)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());

    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));    
  }
  

  public function test_addMentorNeverTooMany()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    $s5 = new StudentR(95);

    $m = new MentorR("blah",array($s,$s2,$s3,$s4));
    
    $this->assertEqual(true,$s5->addMentor($m));
    $this->assertEqual(5,$m->numberOfStudents());
    $this->assertEqual(1,$s5->numberOfMentors());
    
  }
  

  public function test_addStudent()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    $s5 = new StudentR(95);

    $m = new MentorR("blah",array($s,$s2,$s3));

    $this->assertEqual(true,$m->addStudent($s4));
    
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    $this->assertEqual($m,$s4->getMentor(0));    
   
    $this->assertEqual(true,$m->addStudent($s5));
    $this->assertEqual(5,$m->numberOfStudents());
  
  }  


  public function test_addStudentHasExistingMentor()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    $s5 = new StudentR(95);
    $s6 = new StudentR(94);

    $m = new MentorR("blah",array($s,$s2,$s3));
    $m2 = new MentorR("blah2",array($s4,$s5,$s6));

    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
    
    $this->assertEqual($m2,$s4->getMentor(0));
    $this->assertEqual($m2,$s5->getMentor(0));
    $this->assertEqual($m2,$s6->getMentor(0));
    
    $this->assertEqual(true,$m->addStudent($s4));
    
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(3,$m2->numberOfStudents());
    
    $this->assertEqual(2,$s4->numberOfMentors());    
    $this->assertEqual($m2,$s4->getMentor(0));    
    $this->assertEqual($m,$s4->getMentor(1));

    $this->assertEqual(true,$m->addStudent($s5));
  }  


  public function test_removeStudent()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);
    $s5 = new StudentR(95);

    $m = new MentorR("blah",array($s,$s2,$s3,$s5));
    
    $this->assertEqual(false,$m->removeStudent($s4));
    $this->assertEqual(true,$m->removeStudent($s3));
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());   
    $this->assertEqual($m,$s5->getMentor(0));

    $this->assertEqual(false,$m->removeStudent($s2));
    $this->assertEqual(3,$m->numberOfStudents());
  }  
  

  public function test_setStudents_empty()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);

    $m = new MentorR("blah",array($s,$s2,$s3));

    $this->assertEqual(false, $m->setStudents(array()));
    $this->assertEqual(3,$m->numberOfStudents());
  }
  

  public function test_deletStudentDeletesTheMentorIfNotEnoughLeft()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);

    $m = new MentorR("blah",array($s,$s2,$s3));
    $studentP = new ProgramR();
    $mentorP = new ProgramR();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s->delete();
    
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(null,$m->getProgram());
    $this->assertEqual(null,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(0,$s->numberOfMentors());
  }  
  

  public function test_deleteStudentLeavesMentorAloneIfHasEnough()
  {
    $s = new StudentR(99);
    $s2 = new StudentR(98);    
    $s3 = new StudentR(97);
    $s4 = new StudentR(96);

    $m = new MentorR("blah",array($s,$s2,$s3,$s4));
    $studentP = new ProgramR();
    $mentorP = new ProgramR();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s->delete();
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($mentorP,$m->getProgram());
    $this->assertEqual($m,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(0,$s->numberOfMentors());
  } 

  
  
}

File: testbed_ruby/test/associations/many_to_m_star_test.rb
require 'test/unit'
require './src-gen-umple/student_r'
require './src-gen-umple/mentor_r'
require './src-gen-umple/program_r'

module CruiseAssociations
class ManyToMStarTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentR.new(99)
    assert_equal(0,s.number_of_mentors)
  end
  
  def test_ConstructorWithTooFew
    assert_raise RuntimeError do #"Unable to create MentorR, must have at least 3 students"
      s = StudentR.new(99)
      MentorR.new("blah",[s])
    end
  end
  
  
  def test_ConstructorCannotHaveTooMany
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    s5 = StudentR.new(95)
    s6 = StudentR.new(94)
    s7 = StudentR.new(93)
    s8 = StudentR.new(92)
    MentorR.new("blah",[s,s2,s3,s4,s5,s6,s7,s8])
  end
  
  def test_constructorJustBigEnough
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
  
    m = MentorR.new("blah",[s,s2,s3])
  
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
  end
  
  
  def test_constructorMoreThanBigEnough
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
  
    m = MentorR.new("blah",[s,s2,s3,s4])
  
    assert_equal(4,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))
  end  
  
  def test_constructorWatchOutForDuplicateEntries
    assert_raise RuntimeError do #"Unable to create MentorR, must have at least 3 students"
      s = StudentR.new(99)
      MentorR.new("blah",[s,s])
    end
  end
  
  
  def test_constructorExistingMentorsOkay
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    
    m = MentorR.new("blah",[s,s2,s3])
  
    assert_equal(3, m.number_of_students)
    assert_equal(m, s.get_mentor(0))
    assert_equal(m, s2.get_mentor(0))
    assert_equal(m, s3.get_mentor(0))
    assert_equal(0, s4.number_of_mentors)    
  
    m2 = MentorR.new("blah",[s,s2,s4])
    
    assert_equal(3, m.number_of_students)
    assert_equal(m, s.get_mentor(0))
    assert_equal(m, s2.get_mentor(0))
    assert_equal(m, s3.get_mentor(0))
    
    assert_equal(3, m2.number_of_students)
    assert_equal(m2, s.get_mentor(1))
    assert_equal(m2, s2.get_mentor(1))
    assert_equal(m2, s4.get_mentor(0))
  
  end   
  
  
  def test_set_students_tooFew
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    s5 = StudentR.new(95)
  
    m = MentorR.new("blah",[s,s2,s3])
    
    assert_equal(false, m.set_students([s5,s4]))
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
  end
  
  
  def test_set_students_doNotAllowDuplicates
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    
    m = MentorR.new("blah",[s,s2,s3])
    assert_equal(false, m.set_students([s2,s2,s3]))
  end
  
  
  def test_set_students_aboveMinimum
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
  
    m = MentorR.new("blah",[s,s2,s3])
    
    assert_equal(true, m.set_students([s2,s3,s4]))
    assert_equal(3,m.number_of_students)
    assert_equal(0,s.number_of_mentors)
  
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))    
  end
  
  
  def test_add_mentorNeverTooMany
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    s5 = StudentR.new(95)
  
    m = MentorR.new("blah",[s,s2,s3,s4])
    
    assert_equal(true,s5.add_mentor(m))
    assert_equal(5,m.number_of_students)
    assert_equal(1,s5.number_of_mentors)
    
  end
  
  
  def test_add_student
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    s5 = StudentR.new(95)
  
    m = MentorR.new("blah",[s,s2,s3])
  
    assert_equal(true,m.add_student(s4))
    
    assert_equal(4,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    assert_equal(m,s4.get_mentor(0))    
   
    assert_equal(true,m.add_student(s5))
    assert_equal(5,m.number_of_students)
  
  end  
  
  def test_add_studentHasExistingMentor
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    s5 = StudentR.new(95)
    s6 = StudentR.new(94)
  
    m = MentorR.new("blah",[s,s2,s3])
    m2 = MentorR.new("blah2",[s4,s5,s6])
  
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
    
    assert_equal(m2,s4.get_mentor(0))
    assert_equal(m2,s5.get_mentor(0))
    assert_equal(m2,s6.get_mentor(0))
    
    assert_equal(true,m.add_student(s4))
    
    assert_equal(4,m.number_of_students)
    assert_equal(3,m2.number_of_students)
    
    assert_equal(2,s4.number_of_mentors)    
    assert_equal(m2,s4.get_mentor(0))    
    assert_equal(m,s4.get_mentor(1))
  
    assert_equal(true,m.add_student(s5))
  end  
  
  def test_remove_student
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)
    s5 = StudentR.new(95)
  
    m = MentorR.new("blah",[s,s2,s3,s5])
    
    assert_equal(false,m.remove_student(s4))
    assert_equal(true,m.remove_student(s3))
    
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)   
    assert_equal(m,s5.get_mentor(0))
  
    assert_equal(false,m.remove_student(s2))
    assert_equal(3,m.number_of_students)
  end  
  
  
  def test_set_students_empty
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
  
    m = MentorR.new("blah",[s,s2,s3])
  
    assert_equal(false, m.set_students([]))
    assert_equal(3,m.number_of_students)
  end
  
  
  def test_deletStudentDeletesTheMentorIfNotEnoughLeft
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
  
    m = MentorR.new("blah",[s,s2,s3])
    studentP = ProgramR.new
    mentorP = ProgramR.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)
  
    s.delete
    
    assert_equal(0,m.number_of_students)
    assert_equal(nil,m.get_program)
    assert_equal(nil,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(0,s.number_of_mentors)
  end  

  def test_deleteStudentLeavesMentorAloneIfHasEnough
    s = StudentR.new(99)
    s2 = StudentR.new(98)    
    s3 = StudentR.new(97)
    s4 = StudentR.new(96)

    m = MentorR.new("blah",[s,s2,s3,s4])
    studentP = ProgramR.new
    mentorP = ProgramR.new
    
    assert_equal true, s2.set_program(studentP)
    assert_equal true, m.set_program(mentorP)

    s.delete
    
    assert_equal(3,m.number_of_students)
    assert_equal(mentorP,m.get_program)
    assert_equal(m,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(0,s.number_of_mentors)
  end 
  
end
end